common/src/reserved.rs

```diff
@@ -22,55 +22,122 @@ pub struct ReservedState {
+                        .or_insert(voting_power);
+                    Ok::<HashMap<crypto::PublicKey, u64>, String>(map)
+                },
+            )?
+            .into_iter()
+            .collect::<Vec<_>>();
+
+        Ok(validator_set)
     }
 
     pub fn get_governance_set(&self) -> Result<Vec<(PublicKey, VotingPower)>, String> {
-        let mut governance_set = HashMap::new();
-        for member in &self.members {
-            if let Some(delegatee) = &member.governance_delegatee {
-                governance_set
-                    .entry(delegatee.clone())
-                    .and_modify(|v| *v += member.consensus_voting_power)
-                    .or_insert(member.consensus_voting_power);
-            } else {
-                governance_set
-                    .entry(member.name.clone())
-                    .and_modify(|v| *v += member.consensus_voting_power)
-                    .or_insert(member.consensus_voting_power);
-            }
-        }
-        Ok(governance_set
+        let governance_set = self
+            .members
             .iter()
-            .map(|(name, voting_power)| (self.query_public_key(name).unwrap(), *voting_power))
-            .collect())
+            .map(|member| {
+                let name = member
+                    .governance_delegatee
+                    .as_ref()
+                    .map_or_else(|| member.name.as_str(), |name| name.as_str());
+                let public_key = self.query_public_key(&name.to_string()).ok_or_else(|| {
+                    format!("the public key of {name} is not found in the reserved state.")
+                })?;
+                Ok((public_key, member.consensus_voting_power))
+            })
+            .try_fold(
+                HashMap::new(),
+                |mut map, result: Result<(crypto::PublicKey, u64), String>| {
+                    let (public_key, voting_power) = result?;
+                    map.entry(public_key)
+                        .and_modify(|v| *v += voting_power)
+                        .or_insert(voting_power);
+                    Ok::<HashMap<crypto::PublicKey, u64>, String>(map)
+                },
+            )?
+            .into_iter()
+            .collect::<Vec<_>>();
+
+        Ok(governance_set)
     }
 
-    pub fn apply_delegate(&mut self, _tx: &TxDelegate) -> Result<Self, String> {
-        unimplemented!()
+    pub fn apply_delegate(&mut self, tx: &TxDelegate) -> Result<Self, String> {
+        let data = DelegationTransactionData {
+            delegator: tx.data.delegator.clone(),
+            delegatee: tx.data.delegatee.clone(),
+            governance: tx.data.governance,
+            block_height: tx.data.block_height,
+            timestamp: tx.data.timestamp,
+        };
+        if tx.proof.verify(&data).is_err() {
+            return Err("delegation proof verification failed".to_string());
+        }
+        let delegator_name = match self.query_name(&tx.data.delegator.clone()) {
```

Comment:
use `ok_or_else(..)?` instead of `match`.

Reply:
I would refactor like the following code fragment:
```rust
let delegator_name = self
            .query_name(&tx.data.delegator.clone())
            .ok_or_else(|| "delegator does not exist by name".to_string())?;
```